Initial database scheme

DusanBrankovic opened the pull request.

megicivovic commented on travel/src/main/java/com/tx/travel/controller/ExchangeRateController.java:
```diff
@@ -0,0 +1,50 @@
+@CrossOrigin(origins = "*", maxAge = 3600)
+@RestController
+@RequestMapping("api/exchange-rates")
+public class ExchangeRateController {
```
ddi you mean to include all of these files, other than the migration script into this PR? You could use soft rebase to fix this, I can help 😉

DusanBrankovic replied:
I didn't mean to include all of those other files 😅 Thank you for the help offer!

megicivovic commented on travel/src/main/resources/db/migration/V3__dbscheme.sql:
```diff
@@ -0,0 +1,88 @@
+CREATE TABLE IF NOT EXISTS `travel_expense_forms` (
+    `id` VARCHAR(36) NOT NULL,
+
+    `departure_time` TIME NOT NULL,
```
maybe use timestamp here, so it is more standardized?

DusanBrankovic replied:
I totally agree! I will change that throughout the whole scheme.

megicivovic commented on travel/src/main/resources/db/migration/V3__dbscheme.sql:
```diff
@@ -0,0 +1,88 @@
+    `id` VARCHAR(36) NOT NULL,
+    `claim` VARCHAR(255) NOT NULL,
+    `description` VARCHAR(255),
+    `amount` INT DEFAULT 1,
```
does it have to be int?

DusanBrankovic replied:
Thinking about it now, I would put it as unsigned tinyint, since there's no need for negative values. Would that be more fitting?

megicivovic commented on travel/src/main/resources/db/migration/V3__dbscheme.sql:
```diff
@@ -0,0 +1,88 @@
+CREATE TABLE IF NOT EXISTS `travel_authorization_forms` (
+    `id` VARCHAR(36) NOT NULL,
+
+    `departure_time` TIME NOT NULL,
```
also timestamp

megicivovic reviewed (changes requested):
please remove unnecessary  classes and make sure your maven build passes. please have a look at data types with @petkocfc :wink

petkocfc commented on travel/src/main/java/com/tx/travel/model/ExchangeRate.java:
```diff
@@ -0,0 +1,29 @@
+import jdk.jfr.Unsigned;
+import org.hibernate.validator.constraints.Length;
+
+import javax.persistence.*;
```
We should avoid wildcard imports, import only necessary classes 

check: https://www.javadude.com/posts/20040522-import-on-demand-is-evil/

DusanBrankovic replied:
Thank you for the resource! I will change that in the next PR.

petkocfc commented on travel/src/main/java/com/tx/travel/model/ExchangeRate.java:
```diff
@@ -0,0 +1,29 @@
+    private String currency;
+
+    @NotNull
+    private Float amountInRsd;
```
Can we use BigDecimal here?

DusanBrankovic replied:
Thanks for the suggestion, I will look into it and implement changes within the next PR.

DusanBrankovic committed:
Migration scheme excluding Users and Roles tables.

DusanBrankovic committed:
Changed daily_allowances table to more accurately resemble Milena's model.

DusanBrankovic committed:
Updated `exchange_rate` table.
Changed `amount_in_rsd` data type from float to decimal(7,6).
Changed `code` to varchar(3) because of the ISO 4217 standard which states that all currency codes are 3 letters long.

DusanBrankovic committed:
Changed `departure_date` data type from date to timestamp in trips table.
Removed `departure_time` column from travel_authorization_forms and travel_expense_forms since it's not relevant on its own.
Merged `business_trip_end_date` with `business_trip_end_time` and changed its data type to timestamp.

DusanBrankovic committed:
Changed `submission_date` data type from date to timestamp in travel_expense_forms and travel_authorization_forms tables.

DusanBrankovic committed:
Changed expenses table:
Changed `amount` column data type from int to unsigned tinyint since int range is too big for the purpose of the column, and there's no need for negative values.
Renamed `currency` column to `currency_code` to be more descriptive.
Reduced `currency_code` varchar length to 3 (ISO 4217 standard for currency codes)

DusanBrankovic committed:
Changed travel_expense_forms table:
Changed `breakfasts_provided`, `lunches_provided` and `dinners_provided` data type from int(3) to unsigned tinyint because there's no need for negative values and range of the int type is too big for the purpose of the columns.

DusanBrankovic committed:
Changed travel_authorization_forms table:
Renamed `num_breakfast` -> `breakfasts`, `num_lunch` -> `lunches` and `num_dinner` -> `dinners`.
Changed `breakfasts`, `lunches` and `dinners` data type from int(3) to unsigned tinyint because there is no need for negative values and int range is too big for the purpose of these columns.
Changed (*) columns that store decimal values to decimal data type.
Updated `allowance_first_day` and `allowance_last_day` data type from float to unsigned float because there is no need for negative numbers.

(*) daily_allowance_fee, total_allowance, total_allowance_rsd, breakfast_fee, lunch_fee, dinner_fee

DusanBrankovic committed:
Changed trips table:
Removed not null constraint from `tef_id` since it can be null when a trip is in 'In Progress' status.
Increased max length of `country` and `city` columns.

megicivovic commented on travel/src/main/resources/db/migration/V3__dbscheme.sql:
```diff
@@ -0,0 +1,84 @@
+    `id` VARCHAR(36) NOT NULL,
+    `claim` VARCHAR(255) NOT NULL,
+    `description` VARCHAR(255),
+    `amount` TINYINT UNSIGNED DEFAULT 1,
```
isn't this also a decimal number?

DusanBrankovic replied:
Yes! I confused that column's purpose for something else.
I will also change the name to something that makes it clearer as to what data the column is storing.

megicivovic commented on travel/src/main/resources/db/migration/V3__dbscheme.sql:
```diff
@@ -0,0 +1,84 @@
+    `business_trip_end_date` TIMESTAMP NOT NULL,
+    `arrival_date` TIMESTAMP NOT NULL,
+    `accommodation` VARCHAR(255) NOT NULL,
+    `breakfasts` TINYINT UNSIGNED NOT NULL,
```
maybe name this as breakfast_number?

DusanBrankovic replied:
Thanks for the suggestion, I see how this might be incomplete information about what's stored in that column.
I will change it 😄

DusanBrankovic committed:
Code review feedback changes
Reverted data type for `breakfasts_provided`, `lunches_provided` and `dinners_provided` from travel_expense_forms table from tinyint back to int since there can be trips longer than 255 days.
Renamed `claim` to `name` and `amount` to `cost` in expenses table to more accurately describe their use.
Expanded range for columns with decimal data type. (e.g. decimal(7,6) -> decimal(13,6))
Changed travel_authorization_forms table:
Renamed `breakfasts` -> `breakfast_number`, `lunches` -> `lunch_number`, `dinners` -> `dinner_number`. Reverted data types from tinyint to int since there can be trips longer than 255 days.
Renamed `purpose` column to `purpose_of_travel` to be more descriptive.
Renamed `allowance_first_day` -> `first_day_granted_allowance_percentage` and `allowance_last_day` -> `last_day_granted_allowance_percentage`.
Renamed `breakfast_fee` -> `breakfast_fee_percentage`, `lunch_fee` -> `lunch_fee_percentage` and `dinner_fee` -> `dinner_fee_percentage` to more accurately describe stored data in those columns.

petkocfc commented on travel/src/main/resources/db/migration/V3__dbscheme.sql:
```diff
@@ -0,0 +1,84 @@
+
+CREATE TABLE IF NOT EXISTS `travel_expense_forms` (
+    `id` VARCHAR(36) NOT NULL,
+
```
remove newlines

DusanBrankovic committed:
Removed unnecessary empty lines.

DusanBrankovic committed:
Added AUTO_INCREMENT to `id` columns in daily_allowances and cost_centers tables.

petkocfc commented on travel/src/main/resources/db/migration/V3__dbscheme.sql:
```diff
@@ -0,0 +1,81 @@
+
+CREATE TABLE IF NOT EXISTS `travel_expense_forms` (
+    `id` VARCHAR(36) NOT NULL,
+    `business_trip_end_date` TIMESTAMP NOT NULL,
```
do we need this field in this table?

DusanBrankovic replied:
Since we have that same field in 'travel_authorization_forms' table I agree that it's not needed to store same information in two places.
Thank you for pointing it out, I will remove it.

DusanBrankovic replied:
Do we want to store information if user extends their business trip duration in another field or are we updating initially set business_trip_end_date?

petkocfc replied:
I would say we should update initial trip_end_date in that specific scenario

DusanBrankovic replied:
Okay, thank you.

DusanBrankovic committed:
Removed `business_trip_end_date` column from `travel_expense_forms` table because we are storing that same information in `travel_authorization_forms` table.

DusanBrankovic committed:
Renamed `business_trip_end_date` to `trip_end_date` field in `travel_authorization_forms` table.
